### ISSUE TITLE
Initialization discussion

### ISSUE THREAD

**Issue by adbrebs** [bartvm/blocks#384]
Hi all,

First, well done for your hard work! I have just read the tutorial and I have a few remarks/questions. They may sound critical but don't get me wrong, I'm sure that your design choices are well thought out and I just want to understand them.

1. Why do we need to call explicitly brick.initialize? Can't the initialization be done when apply is called? Or during the allocation?

2. When is lazy initialization (isn't the term "lazy configuration" more suitable?) necessary/convenient? The example given in the tutorial (connecting two networks with a linear layer) could be done as easily without the lazy initialization. If it is not necessary/convenient, it may be dangerous to allow too much flexibility especially if Blocks wants to stay quite low-level. Having different ways to do the same thing can be quite confusing for a user and tedious for us to develop (the smaller the library the better).

3. "Many neural network models, especially more complex ones, can be considered hierarchical structures. Even a simple multi-layer perceptron consists of layers, which in turn consist of a linear transformation followed by a non-linear transformation.
As such, bricks can have children. "
The terminology (children, parent) might be slightly confusing. The hierarchy you mention is actually a sequence of bricks in which each brick has one child (the previous brick) and one parent (the next brick). MLP is a superBrick composed of subBricks.

4. Why do we need to call mlp.push_initialization_config when we override the initialization schemes of the children? Would it be possible to remove this step?

5. "By default, all our parameters are set to 0."
It seems dangerous to initialize the parameters to zero when they are allocated. Users might forget to call the initialize method.

I will do a PR with the few typos I have found in the tutorial.

**Comment by rizar** [comment]
Thanks for your feedback!

I hope it is okay that I will answer your question one (or a few) at a time, cause I hate writing too long entries :)

1. In the most general case we might want not to initialize at all. Initialization of a very big model can take quite some time and is not necessary when one loads parameters into the bricks from the disk. We acknowledge, that in most cases it is OK do it automatically and this is a subject of active discussion, see for instance #337. After we switched to initializing parameters with NaNs, (see #378),
one can not simply forget to call `initialize` any more: his cost will be NaN from the beginning. So currently one has to always call `initialize`, which is not very convenient, but neither makes the code much more difficult.

**Comment by rizar** [comment]
5. You are absolutely right, but we do not do it any more! #378 fixes that.

**Comment by rizar** [comment]
(2) Lazy initialization and allocation is almost the only way to deal with the situation when a complex brick A has a few bricks B, C, ... as his components (children). In such cases it is often impossible to fully configure B unless you have C and the other way round. So the Blocks way is:

```python
b = B(partial configuration)
c = C(partial configuration)
a = A(b, c) 
a.initialize()
```

In the second last line `b` and `c` configurations are completed.

A good example of such a brick A is the `AttentionRecurrent` brick that adds an attention mechanism B to a recurrent network C, see it [here](https://github.com/bartvm/blocks/blob/master/blocks/bricks/attention.py#L390)

**Comment by rizar** [comment]
(4) No, because `push_initialization_config` called implicitly by `initialize` in this case will wipe all your changes away. It is important to first propagate the initialization schemes, and then overwrite the results of propagation.

**Comment by rizar** [comment]
(3) I am not sure I fully understand your reasoning, but I would like to emphasize the fact that children not necessarily form a sequence. This was one of PyLearn2 limitations we had to get rid of to make sure that new crazy architectures fit into Blocks nicely. Consider the `Parallel` brick from `blocks/bricks/parallel.py`: his children are applied in parallel.

**Comment by adbrebs** [comment]
Thanks for your answers! #337 answers many of my questions (which are quite similar to those of @fvisin).
1. For loading parameters from the disk, can't we have another initialization scheme class for that?
2,3. Ok I see, thanks
4. Can't the initialization schemes be propagated directly at the creation of the object MLP?
5. ok, I have updated the doc.

**Comment by rizar** [comment]
1. Thanks for this idea, this one has not yet been considered. I would probably not support it, because filling my model with some `NoOpScheme` every time before loading from disk seems rather awkward to me. 

I guess we will stick to explicit initialization at this early stage of Blocks development because it is just the simplest scheme possible which supports all we want to support. An automatic initialization which has certain exceptions would be an additional burden on our shoulders, and we have very limited development power so far. We will most likely promote a two-liner
```python
for brick in model.bricks:
    brick.initialize()
```    
or even `model.initialize_bricks()` soon, just that we have to finish the model argument in #347.

4. Unfortunately not, because they are often unknown at this stage. In terms of bricks A, B, C that I declared above, the brick B does not know his initialization scheme before it becomes a child of the brick A.

In general, many things seems very simple in the MLP case but are much more complicated when you think about more complex architectures. However, we try to keep the MLP case simple enough.

**Comment by adbrebs** [comment]
1)

> because filling my model with some NoOpScheme every time before loading from disk seems rather awkward to me.

It would rather be a `LoadingFromDiskScheme(path_on_disk)`. And the data would be loaded when `initialize()` is called.
Apart from loading the parameters from disk, what else prevent the initialization to be done in `apply`?

5) then can't the default initialization scheme be propagated in the `initialize` method at least? And if you change the initialization scheme of a particular brick beforehand, the default initialization will not be pushed for that particular brick (it will be pushed only to bricks which initialization scheme (you call it `weights_init` I think) is None).

**Comment by bartvm** [comment]
The most important obstacle is mentioned in #337: Currently we don't know whether we have the configuration for initialization available, because we don't know which configuration settings are needed for allocation and which ones for initialization. If we don't know whether we have the configuration available, it's dangerous to try and initialize (it might crash, and even if we catch that, we might have sampled from RNGs, making reproduction of experiments more difficult).

For `weights_init` you could do the propagation more intelligently, but I think there are many other cases where you can't e.g. the parent brick might be intentionally overriding defaults that don't make sense in a context, it might be clear what "not set" is if `None` is an option, etc. So it might be safer to be consistent and clear, and just say that pushing configuration always overrides.

**Comment by rizar** [comment]
As for `LoadingFromDiskScheme`: it would be rather problematic to implement loading by a scheme because the names of different bricks' parameters are not unique in Blocks. It is the hierarchical structure that resolves the ambiguity. Apart from that one might want not load parameters from the disk, but instead takes them from another model trained in the same script. I would not be nice to have a custom scheme for each of such cases.
